File: app/src/main/java/net/squanchy/SquanchyApplication.java
package net.squanchy;

import android.app.Application;
import android.support.annotation.MainThread;

import com.crashlytics.android.Crashlytics;
import com.crashlytics.android.core.CrashlyticsCore;
import com.google.firebase.FirebaseApp;
import com.google.firebase.database.FirebaseDatabase;
import com.twitter.sdk.android.core.Twitter;
import com.twitter.sdk.android.core.TwitterAuthConfig;
import com.twitter.sdk.android.core.TwitterConfig;

import net.danlew.android.joda.JodaTimeAndroid;
import net.squanchy.analytics.Analytics;
import net.squanchy.injection.ApplicationComponent;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import io.fabric.sdk.android.Fabric;
import io.reactivex.android.schedulers.AndroidSchedulers;
import timber.log.Timber;

public class SquanchyApplication extends Application {

    private ApplicationComponent applicationComponent;

    @Override
    public void onCreate() {
        super.onCreate();

        JodaTimeAndroid.init(this);
        setupTracking();
        initializeTwitter();
        initializeFirebase();

        preloadRemoteConfig();
    }

    @SuppressFBWarnings("RV_RETURN_VALUE_IGNORED") // Cannot handle the subscription here
    private void preloadRemoteConfig() {
        applicationComponent().remoteConfig()
                .fetchNow()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        () -> Timber.i("Remote config prefetched"),
                        throwable -> Timber.e(throwable, "Unable to preload the remote config")
                );
    }

    private void setupTracking() {
        initializeFabric();

        Analytics analytics = applicationComponent().analytics();
        analytics.enableExceptionLogging();
        analytics.initializeStaticUserProperties();

        if (BuildConfig.DEBUG) {
            Timber.plant(new Timber.DebugTree());
        }
    }

    private void initializeFabric() {
        CrashlyticsCore crashlyticsCore = new CrashlyticsCore.Builder()
                .disabled(BuildConfig.DEBUG)
                .build();

        Crashlytics crashlytics = new Crashlytics.Builder()
                .core(crashlyticsCore)
                .build();

        Fabric.with(this, crashlytics);
    }

    private void initializeTwitter() {
        TwitterAuthConfig authConfig = new TwitterAuthConfig(
                getString(R.string.api_value_twitter_api_key),
                getString(R.string.api_value_twitter_secret)
        );

        TwitterConfig twitterConfig = new TwitterConfig.Builder(this)
                .twitterAuthConfig(authConfig)
                // TODO .logger(new TimberLogger())
                .debug(BuildConfig.DEBUG)
                .build();

        Twitter.initialize(twitterConfig);
    }

    private void initializeFirebase() {
        // Calling initializeApp() should not be necessary, because Firebase uses the ContentProvider hack
        // (see FirebaseInitProvider). That means that, given we are not running multiple processes, we should
        // not need this. Unfortunately that's not the case; as of Firebase 10.2.0, the app sometimes crashes.
        if (FirebaseApp.initializeApp(this) == null) {
            Timber.e(new IllegalStateException("Initializing Firebase failed"));
        }
        FirebaseDatabase.getInstance().setPersistenceEnabled(true);
    }

    @MainThread
    public ApplicationComponent applicationComponent() {
        if (applicationComponent == null) {
            applicationComponent = ApplicationComponent.Factory.create(this);
        }

        return applicationComponent;
    }
}

File: app/src/main/java/net/squanchy/SquanchyApplication.kt
package net.squanchy

import android.app.Application
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.core.CrashlyticsCore
import com.google.firebase.database.FirebaseDatabase
import io.fabric.sdk.android.Fabric
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import io.reactivex.rxkotlin.subscribeBy
import net.danlew.android.joda.JodaTimeAndroid
import net.squanchy.injection.createApplicationComponent
import timber.log.Timber

class SquanchyApplication : Application() {

    val applicationComponent by lazy { createApplicationComponent(this) }

    private val subscriptions = CompositeDisposable()

    override fun onCreate() {
        super.onCreate()

        JodaTimeAndroid.init(this)
        setupTracking()
        initializeFirebase()
    }

    private fun setupTracking() {
        initializeFabric()

        with(applicationComponent.analytics()) {
            enableExceptionLogging()
            initializeStaticUserProperties()
            trackFirstStartUserNotLoggedIn()
            trackFirstStartNotificationsEnabled()
        }

        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
        }
    }

    private fun initializeFabric() {
        val crashlyticsCore = CrashlyticsCore.Builder()
            .disabled(BuildConfig.DEBUG)
            .build()

        val crashlytics = Crashlytics.Builder()
            .core(crashlyticsCore)
            .build()

        Fabric.with(this, crashlytics)
    }

    private fun initializeFirebase() {
        FirebaseDatabase.getInstance().setPersistenceEnabled(true)
        preloadRemoteConfig()
        preloadTracks()
    }

    private fun preloadRemoteConfig() {
        subscriptions += applicationComponent.remoteConfig()
            .fetchNow()
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeBy(
                onComplete = { Timber.i("Remote config prefetched") },
                onError = { throwable -> Timber.e(throwable, "Unable to preload the remote config") }
            )
    }

    private fun preloadTracks() {
        subscriptions += applicationComponent.tracksRepository()
            .tracks()
            .subscribeBy(
                onNext = { Timber.d("Tracks prefetched: $it") },
                onError = { throwable -> Timber.e(throwable, "Unable to preload the tracks") }
            )
    }

    override fun onTerminate() {
        super.onTerminate()
        // Best-effort to clear any outstanding subscriptions — mostly for RxLint's sake, as this method is not
        // guaranteed to be called (and, in fact, it's almost never). But that's ok as those subscription will
        // have long terminated (and unsubscribed) by the time the process is killed, and even if they hadn't,
        // we'd be killing the process anyway, which is the most extreme form of unsubscription ;)
        subscriptions.clear()
    }
}
